[react-dom] Refactor event priority handling to its own module

trueadm opened the pull request:
The team have mentioned a bunch of times that it felt wrong that DOM Event Priority handling was all coupled with the SimpleEventPlugin, especially as might want to extend on the priorities with newer features and refactors in the future.

This PR breaks out the priority logic into a dedicated module and makes uses of Maps for faster lookups (I benchmarked before and after and Maps were consistently faster than object property lookups in Chrome and Safari, neutral in Firefox). We should also get a slight code size saving by some of the code golfing that I did whilst refactoring the code.

philipp-spiess commented on packages/react-dom/src/events/DOMEventPriority.js:
```diff
@@ -0,0 +1,117 @@
+import * as DOMTopLevelEventTypes from './DOMTopLevelEventTypes';
+import {DiscreteEvent, UserBlockingEvent, ContinuousEvent} from 'shared/ReactTypes';
+
+// We use a Set here rather than an Array or Object map.
```
Out of curiosity: Have you also tried a switch statement?

trueadm replied:
I did and it was even faster, but not by much. The issue with switches was that it was also quite a bit more code size comared to before.

trueadm commented:
I did some benchmarking on FF and Safari and it looks like using multiple Sets is not as optimal as Chrome's V8. I've changed my approach to use Maps and moved things around so we don't do multiple lookups – this is now consistently faster in Chrome and Safari, neutral in Firefox.

necolas commented on packages/react-dom/src/events/DOMEventProperties.js:
```diff
@@ -0,0 +1,175 @@
+  UserBlockingEvent,
+  ContinuousEvent,
+} from 'shared/ReactTypes';
+import {simpleEventPluiginEventTypes} from './SimpleEventPlugin';
```
Typo: simpleEvent**Pluigin**EventTypes

But this file now imports the `SimpleEventPlugin` and creates a cyclic dependency between that and `DOMEventProperties`

philipp-spiess commented on packages/react-dom/src/events/DOMEventProperties.js:
```diff
@@ -0,0 +1,184 @@
+  priority: EventPriority,
+): void {
+  // As the event types are in pairs of two, we need to iterate
+  // through in twos.
```
😮 neat!

necolas replied:
eek though

necolas commented on packages/react-dom/src/events/DOMEventProperties.js:
```diff
@@ -0,0 +1,189 @@
+// here once. If we remove or refactor the
+// SimpleEventPlugin, we should also remove or
+// update the below line.
+export const simpleEventPluiginEventTypes = {};
```
Typo here, should be "Plugin"
